frontend/projects/upgrade/src/app/core/stratification-factors/store/stratification-factors.effects.ts

```diff
@@ -0,0 +1,105 @@
+import { Injectable } from '@angular/core';
+import { Router } from '@angular/router';
+import { Actions, createEffect, ofType } from '@ngrx/effects';
+import { map, switchMap, filter, catchError, withLatestFrom } from 'rxjs/operators';
+import { Store, select } from '@ngrx/store';
+import { AppState } from '../../core.state';
+import * as StratificationFactorsActions from './stratification-factors.actions';
+import { StratificationFactorsDataService } from '../stratification-factors.data.service';
+import { selectAllStratificationFactors } from './stratification-factors.selectors';
+
+@Injectable()
+export class StratificationFactorsEffects {
+  constructor(
+    private actions$: Actions,
+    private router: Router,
+    private store$: Store<AppState>,
+    private stratificationFactorsDataService: StratificationFactorsDataService
+  ) {}
+
+  fetchStratificationFactors$ = createEffect(() =>
+    this.actions$.pipe(
+      ofType(StratificationFactorsActions.actionFetchStratificationFactors),
+      withLatestFrom(this.store$.pipe(select(selectAllStratificationFactors))),
+      switchMap(() =>
+        this.stratificationFactorsDataService.fetchStratificationFactors().pipe(
+          map((data: any) =>
+            StratificationFactorsActions.actionFetchStratificationFactorsSuccess({
+              stratificationFactors: data,
+            })
+          ),
+          catchError(() => [StratificationFactorsActions.actionFetchStratificationFactorsFailure()])
+        )
+      )
+    )
+  );
+
+  deleteStratificationFactor$ = createEffect(() =>
+    this.actions$.pipe(
+      ofType(StratificationFactorsActions.actionDeleteStratificationFactor),
+      map((action) => action.factor),
+      filter((factor) => !!factor),
+      switchMap((factor) =>
+        this.stratificationFactorsDataService.deleteStratificationFactor(factor).pipe(
+          map((data: any) => {
+            this.router.navigate(['/participants']);
+            const stratificationFactor = {
+              ...data[0],
+              factor: data[0].stratificationFactorName,
```

Comment:
we're accessing a property on a potentially undefined or null object. TypeScript surprisingly won't throw any warning squiggles for array indices that are potentially undefined (https://github.com/microsoft/TypeScript/issues/13778), so we have to be careful.

I think it's probably unlikely that we'd get something undefined here, but we have  also defined this data object as `any` also, so we're flying a little blind with type safety in this function.

Reply:
So actually we are returning an array of object from the backend and `data[0]` refers to the first object, which is the deleted row object. I don't think this can be a null object, still if you think, we can add a conditional check. If we have `data.stratificationFactorName`, then `stratificationFactorName` wont exist on object type.
<img width="842" alt="Screenshot 2023-10-19 at 11 08 11 AM" src="https://github.com/CarnegieLearningWeb/UpGrade/assets/33730817/f9c374a8-f844-4d66-bfa6-223014d3c387">

Reply:
@danoswaltCL I have added the conditional check for this.